### DISCUSSION TITLE
Use || instead | for pattern matching 'or' patterns to free up | for matching union types

### DISCUSSION THREAD

**Discussion by therealcisse** [scala/scala3#16587] [Compiler internals Q&A] [answered]
Scala 3 has union types but unfortunately they can't be used in pattern matching since `|` is already used for 'or' patterns. There have been some unsuccessful efforts to change this.

By using  `||` for 'or' pattern, `|` can be used for pattern matching union types.

```scala
final case class Cat(name: String)
final case class Dog(name: String)

val animal = Cat(name = "Puffy")

// or pattern matching
animal match {
  case Cat(name) || Dog(name) => println(s"cat or dog: $name")
}


// Union pattern matching
animal match {
  case a: Cat | Dog => println(s"cat or dog: ${a.name}")
}

````

This was inspired by dart 3. See https://github.com/dart-lang/language/blob/master/accepted/future-releases/0546-patterns/feature-specification.md#logical-or-pattern

Naturally this might require considerable migration. But still worth taking a look at since union types matching could be very useful.

**Comment by sjrd** [comment]
You can do
```scala
case a: Cat | Dog =>
```

**Reply by therealcisse** [reply]
There seems to be some discrepancies:

The above doesn't compile although that seems unintentional: https://scastie.scala-lang.org/DKhAorHkS0KgYUJwF6PjyQ

Looks like an overlapping variation of both ways is working. Would be nice to separate them into different keywords.

**Reply by sjrd** [reply, marked as answer]
Sorry, got my syntax wrong. I meant
```scala
case a: (Cat | Dog) =>
```
